controllers/installation_controller.go

```diff
@@ -189,6 +197,207 @@ func (r *InstallationReconciler) ReportInstallationChanges(ctx context.Context,
+		return fmt.Errorf("failed to get version metadata configmap: %w", err)
+	}
+
+	// pull the artifact from the artifact location pointed by EmbeddedClusterMetadata. This property
+	// points to a repository inside the registry running on the cluster.
+	location, err := artifacts.Pull(ctx, log, r.Client, in.Spec.Artifacts.EmbeddedClusterMetadata)
+	if err != nil {
+		return fmt.Errorf("failed to pull version metadata: %w", err)
+	}
+	defer os.RemoveAll(location)
+
+	// now that we have the metadata locally we can reads its information and create the config map.
+	fpath := filepath.Join(location, "version-metadata.json")
+	data, err := os.ReadFile(fpath)
+	if err != nil {
+		return fmt.Errorf("failed to read version metadata: %w", err)
+	}
+
+	cm.Name = nsn.Name
+	cm.Namespace = nsn.Namespace
+	cm.Data = map[string]string{"metadata.json": string(data)}
+	if err := r.Create(ctx, &cm); err != nil {
+		return fmt.Errorf("failed to create version metadata configmap: %w", err)
+	}
+	return nil
+}
+
+// CopyArtifactsToNodes copies the installation artifacts to the nodes in the cluster.
+// This is done by creating a job for each node in the cluster, which will pull the
+// artifacts from the internal registry.
+func (r *InstallationReconciler) CopyArtifactsToNodes(ctx context.Context, in *v1beta1.Installation) (bool, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Evaluating jobs for copying artifacts to nodes", "installation", in.Name)
+	if in.Spec.Artifacts == nil {
+		in.Status.State = v1beta1.InstallationStateFailed
+		in.Status.Reason = "Artifacts locations not specified for an airgap installation"
+		return false, nil
+	}
+
+	var nodes corev1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		return false, fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	status := map[string]string{}
+	ready := true
+	for _, node := range nodes.Items {
+		log.Info("Evaluating job for node", "node", node.Name)
+		nsn := types.NamespacedName{
+			Name:      fmt.Sprintf("copy-artifacts-%s", node.Name),
+			Namespace: "embedded-cluster",
+		}
+
+		// we start by assuming that the jobs has finished successfuly.
+		status[node.Name] = "JobSucceeded"
+
+		// we first verify if a job already exists for the node, if not then one is
+		// created and we move to the next node.
+		var job batchv1.Job
+		if err := r.Get(ctx, nsn, &job); err != nil {
+			if !errors.IsNotFound(err) {
+				return false, fmt.Errorf("failed to get job: %w", err)
+			}
+			ready = false
+			status[node.Name] = "JobCreated"
+			log.Info("No job for node found", "node", node.Name)
+			if err := r.CreateArtifactJobForNode(ctx, in, node); err != nil {
+				return false, fmt.Errorf("failed to create job for node: %w", err)
+			}
+			continue
+		}
+
+		// generate a hash of the current config so we can detect config changes.
+		hash, err := r.HashForAirgapConfig(in)
+		if err != nil {
+			return false, fmt.Errorf("failed to hash airgap config: %w", err)
+		}
```

Comment:
I feel like this doesn't need to be in the loop

Reply:
Good point. Addressed.